internal/client.go

```diff
@@ -643,9 +644,18 @@ type (
 		// Optional: defaulted to Fail.
 		WorkflowIDConflictPolicy enumspb.WorkflowIdConflictPolicy
 
+		// WorkflowOperation - Operation to execute at Workflow Start. For example, see PrepareUpdateWorkflowOperation
+		// to perform Update-with-Start. Note that if the workflow is already running and WorkflowIDConflictPolicy is
+		// set to UseExisting, the start is skipped and only the operation is executed.
+		//
+		// Optional: defaults to nil.
+		WorkflowOperation StartWorkflowOperation
```

Comment:
Initially I wanted to make this a list of operations, but I think having a single operation is just simpler and will work for most use cases. If we wanted to support multiple operations, we can introduce a "composite operation" type which can take a list of operations.

Reply:
Also, the name was chosen to allow more than just Update operations in the future. For instance a `SignalWorkflowOperation`.

Reply:
> And I think we should consider doing signal operation right now and deprecating signal with start.

You mean as part of this PR?

Reply:
> You mean as part of this PR?

Nah, but soon IMO.

> What's your reasoning?

Because in general the idea of having additional operations on a workflow makes more sense than a single additional operation. Users of our API (now and HTTP in the future) are lucky enough to have this representation, we should too. The same reasons it is `repeated` in proto are the reasons for making it a slice here. We could have done composite there too, why did we not?

Reply:
+1, makes sense for this to be a list I think.

Reply:
I'm with Stephan on adding a composite operation later in the relatively small chance we build that.
* That's "gracefully revealed complexity" which makes "easy things easy and hard things possible." 
* I don't want to mislead developers into thinking it's supported and waste their time integrating only to learn via runtime exceptions via testing that it doesn't work.
* There's no real product interest in supporting a list of operations for the time being, and it's not in danger of being added to our roadmap even in 2025.

Reply:
On further thought I agree. There's more types of future invalid combinations we can prevent by encoding composition into the type system (more so in not-Golang, but, still).

However I agree the mismatch with the API is weird. Ideally we'd change the API.

Reply:
Re: supporting signal-with-start, it might be worth doing upfront just to validate that our design is properly able to handle both cases.  But haven't thought it through and up to you.

Reply:
:+1: I personally like the collection (in all APIs), but don't have a strong opinion, this is fine too. I do think we should consider our HTTP/gRPC API user facing and we should have it as similar as the SDK APIs, but meh.

I do still also prefer the term "additional" operations. But even if that is also not what y'all want, I would recommend at least removing the word `Workflow` from the field name since it's in workflow options, but don't have a strong opinion there either.

Reply:
> And I think we should consider doing signal operation right now and deprecating signal with start.

Until the server's MultiOperation API supports Signal, we could do this right now:

```
func (wc *WorkflowClient) ExecuteWorkflow(ctx context.Context, options StartWorkflowOptions, workflow interface{}, args ...interface{}) (WorkflowRun, error) {
	// TODO: use MultiOperation API once there is server-side support
	if op, ok := options.WithStartOperation.(*SignalWorkflowOperation); ok {
		return wc.SignalWithStartWorkflow(ctx, options.ID, op.signalName, op.arg, options, workflow, args...)
	}
```

The two downsides I'm seeing are (1) anyone using interceptors might be surprised and (2) error messages might change once we move to the MultiOperation API (behavior should be identical AFAICT).

Reply:
Yes, my only concern was consistency from user POV. A user should be able to signal with start, update with start, signal + update with start, etc the same way. We don't have to get there in this PR, just a general thought about proper consistent APIs.

Reply:
FWIW, I prototyped the Signal work and it could look like this:

```go
        sig, _ := client.NewSignalWorkflowOperation("SIGNAL", 1)
	wfHandle, err := c.ExecuteWorkflow(
		context.Background(),
		client.StartWorkflowOptions{
			ID:                       wfID,
			TaskQueue:                taskQueue,
			WithStartOperation:       sig,
			WorkflowIDConflictPolicy: enumspb.WORKFLOW_ID_CONFLICT_POLICY_USE_EXISTING,
		},
		Workflow,
	)
```

Reply:
Not the biggest fan of "WithStart" over "Additional" and ideally the top-level `NewSignalWorkflowOperation` would be named to make it clear it's only for this use case and not general purpose.